### ISSUE TITLE
[FR] Add fuzzy mode

### ISSUE THREAD

**Issue by NightMachinary** [Genivia/ugrep#31]
I think this shouldn't be too hard? See [this](https://oremacs.com/swiper/#ivy--regex-fuzzy) for turning the text into a fuzzy search and [this](https://github.com/lewang/flx) for sorting them intelligently. I think `ugrep` beats `fzf` in searching file's content interactively.

**Comment by genivia-inc** [comment]
Coincidentally I have been working on that by extending the RE/flex matcher with a FuzzyMatcher subclass. I haven't worked out all of the details yet, so I cannot 100% promise this will be included very soon. It would be nice to use the short option `-Z` (`--fuzzy`) as this option is rarely used in practice we may assume (with GNU/BSD grep), but that would break compatibility with GNU grep, unfortunately. Still, I like that idea. Also, `-Z` could accept a parameter to control the error threshold.

**Comment by genivia-inc** [comment]
I've build a FuzzyMatcher pattern matcher. This is more powerful that the simplistic insertion of `.*` between characters. The FuzzyMatcher handles things like this:

pattern | max | matches                           | but not
------- | --- | --------------------------------- | ---------------------------
`abc`   | 1   | `abc`, `ab`, `ac`, `axc`, `axbc`  | `a`, `axx`, `bc`
`año`   | 1   | `año`, `ano`, `ao`                | `anno`
`ab_cd` | 2   | `ab_cd`, `ab-cd`, `ab Cd`, `abCd` | `ab\ncd`, `Ab_cd`, `Abcd`
`a\d+z` | 1   | `a1z`, `a123z`, `az`, `axz`       | `axxz`, `A123z`, `123z`

See https://github.com/Genivia/FuzzyMatcher for details.

After some more testing and improvements I'll integrate this in ugrep.

**Comment by genivia-inc** [comment]
One practical question that comes up is what short option to use for fuzzy matching. Earlier I had suggested `-Z`, but that is already taken and should be preserved for compatibility with GNU grep. Another choice could be `-Y`, replacing the existing option `-Y`, `--empty` (permit empty matches). I think that is a more attractive choice, because I doubt that `-Y` is used a lot and we will still keep its long `--empty` alternative. This means that we might have the following new option `-Y` with an optional `MAX` argument:

    -Y[MAX], --fuzzy[=MAX]
            Fuzzy mode: report approximate pattern matches with MAX errors.  A
            character deletion, insertion or substitution counts as one error.
            The default MAX is 1.  No whitespace may be given between -Y and
            its argument MAX.

I figured that a default value of 1 might be reasonable, certainly for short patterns.

By comparison, note that the classic [agrep utility](https://linux.die.net/man/1/agrep) requires a `NUM` max errors specified with `-E NUM` (`--max-errors=NUM`) or simply as `-1`, `-2` etc which is nice. However, we already reserved `-1`, `-2` etc to limit the recursion depth so that is out of the question I suppose. Note that ugrep has many more features compared to agrep and adding approximate matching to ugrep makes ugrep an attractive alternative IMO.

Option `-Y` does not yet factor in the possibility to order matches by relevance. Ordering the results by relevance will require two passes over each file. The first pass determines the relevance as a numeric value. A second pass then reports exact or approximate matches. By comparison, also agrep makes two passes for option `-B, --best-match`.

**Comment by NightMachinary** [comment]
I personally try to always use long form of options and write wrapper functions for stuff I use a lot, so I don’t particularly care. It’s a good idea to give common options non-capital, mnemonic short forms, but given the constraints here I doubt a better option is possible.

Ordering the results is definitely needed for fuzzy search. Having it as a second option is enough though (again, I don’t care about these stuff because of wrappers, but I imagine it would inconvenience others). If there is a “sort” option, integrating this BestMatch feature there makes it easier to find. Otherwise, sth like --sort-best-match is better (again, easier to find in the man page.). My suggestion on the two-pass sort option is to make the common case the default; My intuition is that people will want the two-pass on files, but not on stdin. (Assuming the stdin supports streaming.)

**Comment by genivia-inc** [comment]
> Ordering the results is definitely needed for fuzzy search. Having it as a second option is enough though (again, I don’t care about these stuff because of wrappers, but I imagine it would inconvenience others). If there is a “sort” option, integrating this BestMatch feature there makes it easier to find.

Your suggestion make sense. There is a `--sort` option. I'm thinking that `--sort=best` and `--sort=rbest` should order matching files by most relevant to least relevant. This will require a first pass over a file.

I will add `--sort=best` in a later update when I have time, after adding `-Z` (`--fuzzy`) in the upcoming update. Yes, I decided that `-Z` would probably be best. BSD grep gives `-Z` a different meaning than GNU grep, so it's not really reserved for a specific purpose that users are relying on. Rather, `--null` can be used anyway.

> My suggestion on the two-pass sort option is to make the common case the default; My intuition is that people will want the two-pass on files, but not on stdin. (Assuming the stdin supports streaming.)

Yes, a second pass is required to produce results ordered by relevance. If stdin is not seekable then we will need to exclude it from the results or store it to search repeatedly, like `-Q` already does.

**Comment by genivia-inc** [comment]
I've added `--sort=best` to the search options. When combined with option `-Z`, this performs an initial search of each file to determine the edit distance cost on which the best matches are sorted by the worker threads.

Note:
1. The initial search is cheap if the first matches at the begin of the file are exact matches, since the first pass does not need to search the whole file in that case.
2. If there is no fuzzy match found during the initial search, then this is cheap too because we no longer need to consider this file.
3. Archive (zip, tar.gz etc) contents are not sorted, but fuzzy matches in archives are recorded and the archives are listed in best match order.
4. Standard input is not sorted, but shown first or part of the command line argument files and directories, which are always shown in the same order as specified on the command line. This is similar to other unix utilities, e.g. `ls`.
5. The new `--format` field `%Z` shows the edit distance cost of a fuzzy match.

I wonder if or how close we can get to produce fzf-like output or if fzf could be improved by using ugrep as search tool. Right now, ugrep has fuzzy search options to simulate agrep to a large extent, but not fzf. And maybe that is OK. I understand that fzf is an entirely different tool that is used for different but overlapping workflow tasks.

I closed this issue for now, but the discussion should perhaps continue to figure that last part out.

**Comment by NightMachinary** [comment]
@genivia-inc I have taken a look at the new fuzzy matcher, and while it's an excellent addition to `ugrep` (Thanks!), but it does not address my usecase. 😅 (It seems to me that it is only useful for typos.)

I want an fzf-like matcher, so I can easily search my notes and code to find what I need. For example, I remember that to I have some code to draw a plot in one of my projects, and I want to find this code because I have forgotten how to draw a plot. I remember that the line that did the work contained `dataframe` and `lambda`. In fzf, I can just use the query `dataframe lambda`, and it'll find the line. Doing this with regexes is a lot harder in the general case (because of the unknown order of the terms, the terms being incomplete, e.g., what I remember as `dataframe` is `my-dataframe` in the code), and if I use `dataframe.*lambda`, it'll highlight the junk `.*`, which makes it hard to find what I am looking for in the search results.

**Comment by genivia-inc** [comment]
I know that fzf uses several conventions for patterns that differ from regex patterns.

> what I remember as dataframe might actually be my-dataframe), and if I use dataframe.*lambda, it'll highlight the junk .*, which makes it hard to find what I am looking for in the search results.

That can still be done with regexes, but we need an "and" operator instead of filling the pattern with `.*`. That also allows for any order of the patterns on a matching line. Note that an "or" operator is trivial with pattern `|` or with options `-e` to specify multiple words/patterns.

IMHO, fuzzy matching done by fzf only allows character insertions, which is a poor man's implementation of fuzzy matching/search and has limited applicability. In fact, [fuzzy matching](https://en.wikipedia.org/wiki/Fuzzy_matching_(computer-assisted_translation)) uses approximate [string matching](https://en.wikipedia.org/wiki/Approximate_string_matching) that is a well-known problem and requires algorithms to support string edits. True fuzzy matching/search is driven by the notion of "edit distance", which is more useful, such as character substitutions in matches. For example, if you search for `my-dataframe` it will match `my_dataframe` and `mydataframe` or even `MyDataFrame` with option `-i` (case-insensitive match).

I've thought about adding "and" and "not" operators as options `--and PATTERN` and `--not PATTERN`, to select lines when the specified additional constraints apply. I may add these when I have time.

It would also be useful to add new features to the query UI, such as to navigate the file system by selecting a directory to `cd` into and to `cd` out of directories.

PS. as we're talking about `--not`, note that `-N PATTERN` is different than a `--not`, because it excludes matching patterns by skipping them, e.g. searching for any number `\d+' except in strings is possible with `ugrep '\d+' -N '".*?"'`, whereas `'d+` `--not '".*?"` would match a line with 123 when it does not contain a string, missing the case where 123 is not within the string.

**Comment by NightMachinary** [comment]
[**@genivia-inc**](https://github.com/genivia-inc) commented on [Jul 23, 2020, 9:21 PM GMT+4:30](https://github.com/Genivia/ugrep/issues/31#issuecomment-663115947 "2020-07-23T16:51:21Z - Replied by Github Reply Comments"):
> I know that fzf uses several conventions for patterns that differ from regex patterns.
> 
> > what I remember as dataframe might actually be my-dataframe), and if I use dataframe._lambda, it'll highlight the junk ._, which makes it hard to find what I am looking for in the search results.
> 
> That can still be done with regexes, but we need an "and" operator instead of filling the pattern with `.*`. That also allows for any order of the patterns on a matching line. Note that an "or" operator is trivial with pattern `|` or with options `-e` to specify multiple words/patterns.
> 
> IMHO, fuzzy matching done by fzf only allows character insertions, which is a poor man's implementation of fuzzy matching/search and has limited applicability. In fact, [fuzzy matching](https://en.wikipedia.org/wiki/Fuzzy_matching_(computer-assisted_translation)) uses approximate [string matching](https://en.wikipedia.org/wiki/Approximate_string_matching) that is a well-known problem and requires algorithms to support string edits. True fuzzy matching/search is driven by the notion of "edit distance", which is more useful, such as character substitutions in matches. For example, if you search for `my-dataframe` it will match `my_dataframe` and `mydataframe` or even `MyDataFrame` with option `-i` (case-insensitive match).
> 
> I've thought about adding "and" and "not" operators as options `--and PATTERN` and `--not PATTERN`, to select lines when the specified additional constraints apply. I may add these when I have time.
> 
> It would also be useful to add new features to the query UI, such as to navigate the file system by selecting a directory to `cd` into and to `cd` out of directories.
> 
> PS. as we're talking about `--not`, note that `-N PATTERN` is different than a `--not`, because it excludes matching patterns by skipping them, e.g. searching for any number `\d+' except in strings is possible with` ugrep '\\d+' -N '"._?"'`, whereas` 'd+  \--not '"._?"\` would match a line with 123 when it does not contain a string, missing the case where 123 is not within the string.

I had not noticed that the fuzzy mode works with regexps. That certainly increases its usefulness!

I think this `and` operator will solve my usecase, but only if it is integrated into the interactive query mode. As far as I understand, `fzf` is using ` ` (whitespace) as this `and` operator, and this is a very natural, intuitive choice. Perhaps you can add a `--and-operator="STR"` that makes `STR` act as the `and` operator in the query mode. I think a more thorough look at the syntax conventions of `fzf` might also bear useful lessons. I personally have not used complex `fzf` patterns, but my impression is that it supports a good query syntax for them.

A `not` operator like `!` will also be nice.

So to sum up my thoughts:

- ` ` becomes the `and` operator (optionally allow customization by the user, not really necessary)
- ` ` also acts as a grouping construct, e.g., `world ^hell` becomes `(world)AND(^hell)`.
- `!` becomes the `not` operator
- `\ ` and `\!` mean their original character

**Comment by genivia-inc** [comment]
Very good. Thank you for your thoughtful suggestions!

There are several directions we can take to implement the "and" and "not" operators. Either by escaping them, e.g. as `\&` and `\!` which will not interfere with the current regex syntax and usage, or as you point out, use space and `!` which means that escapes will be needed for their original meaning. I can see pro/cons for both.

**Comment by NightMachinary** [comment]
[**@genivia-inc**](https://github.com/genivia-inc) commented on [Jul 23, 2020, 11:58 PM GMT+4:30](https://github.com/Genivia/ugrep/issues/31#issuecomment-663190502 "2020-07-23T19:28:07Z - Replied by Github Reply Comments"):
> Very good. Thank you for your thoughtful suggestions!
> 
> There are several directions we can take to implement the "and" and "not" operators. Either by escaping them, e.g. as `\&` and `\!` which will not interfere with the current regex syntax and usage, or as you point out, use space and `!` which means that escapes will be needed for their original meaning. I can see pro/cons for both.

Another option is to have the unescaped versions, but only if the user supplies a flag `--regex-extensions` or `--nonstandard-regex` or sth. I think this might be the better choice, because having escaped characters being magic is non-intuitive, and harder to type, too (considering that presumably using the operators is more common than the actual characters.). Same typing consideration for `&` vs the spacebar; One rarely needs to search for the space character (I personally almost always use `\s` in the rare times I want to include whitespace), but using the AND operator is very common and I use an average of one in my `fzf` queries.

**Comment by genivia-inc** [comment]
A new `--not` option is now available since v2.5.1 to ignore lines that match the specified not-match-patterns.

A `--not -e NOTPATTERN` is essentially the same as `-N '.*NOTPATTERN.*'` to ignore and "hide" all `.*NOTPATTERN.*` matches from the search, i.e. ignoring lines that contain `NOTPATTERN`, but optimized for speed. I may further optimize this for speed at a later stage using a different approach.

We still need an `--and` option to cover your suggested use case with patterns that have an implicit "and" through spacing by introducing a new option for compact "and", "or" and "not" regex patterns. For example `PATT1 PATT2` matches lines that contain both `PATT1` and `PATT2` and `PATT1 !PATT2` matches lines that contain `PATT1` but not `PATT2`. I will work on that further when I have time.

Because regex patterns are specified to search, not strings, brackets are allowed to group sub patterns. This poses a design challenge. Consider for example `(PATT1 PATT2)|PATT3`. This cannot be translated properly to the option `--and` with patterns directly, at least not without rearranging the regex pattern to produce the conjunctive normal form `-e 'PATT1|PATT3' --and -e 'PATT2|PATT3'`. We could do that, but that seems an overkill. IMO we should only allow spacing to be interpreted as "and" outside of bracketed expressions.

As for the "not" operator, perhaps we should consider something else instead of `!`:
1. when a `!` is used on the command line with csh and tcsh, it is always expanded as per history mechanism, even when used in strings and it isn't easy to escape. That can be very frustrating for users. Bash has no issues with this.
2. a `^` is normally an anchor. This anchor is only useful at the start of a regex pattern. Because the "not" patterns always follow an initial non-empty pattern to search, using `^` after a space to indicate "not" is viable, e.g. `PATT1 ^PATT2`.
3. or `-` could be used (after a space) to indicate "not", e.g. `PATT1 -PATT2`.
4. with 3) above, the "and" could also be formed with a `+` like so: `PATT1 +PATT2`. This allows a space to still fuzzy-match a space, i.e. match the space or something else.

**Comment by NightMachinary** [comment]
[**@genivia-inc**](https://github.com/genivia-inc) commented on [Aug 19, 2020, 6:51 PM GMT+4:30](https://github.com/Genivia/ugrep/issues/31#issuecomment-676438574 "2020-08-19T14:21:44Z - Replied by Github Reply Comments"):

Thanks!

> Because regex patterns are specified to search, not strings, brackets are allowed to group sub patterns. This poses a design challenge. Consider for example `(PATT1 PATT2)|PATT3`. This cannot be translated properly to the option `--and` with patterns directly, at least not without rearranging the regex pattern to produce the conjunctive normal form `-e 'PATT1|PATT3' --and -e 'PATT2|PATT3'`. We could do that, but that seems an overkill. IMO we should only allow spacing to be interpreted as "and" outside of bracketed expressions.

I think it'd be better to do it the overkill way. I [vaguely remember](https://en.wikipedia.org/wiki/Conjunctive_normal_form) that there are good algorithms for turning an algebraic expression into a normalized form, and this would be an intuitive feature to have. A possible usecase: `(\(\) {)|(alias =) SOMETHING` to search for the definition of a shell function or alias. It's very natural to compose regexes this way, rather having to manually normalize the patterns into something the raw regex understands.

> 
> As for the "not" operator, perhaps we should consider something else instead of `!`:
> 
> 1.  when a `!` is used on the command line with csh and tcsh, it is always expanded as per history mechanism, even when used in strings and it isn't easy to escape. That can be very frustrating for users. Bash has no issues with this.

I use zsh, and zsh has the `!` history thing, but it isn't hard to escape. I just use the single quotes: `echo 'Hello!'`. Are those  shells different? I personally would use the single quotes for regexes regardless, as a lot of their syntax is special in zsh (`*()?{}...`).

> 2.  a `^` is normally an anchor. This anchor is only useful at the start of a regex pattern. Because the "not" patterns always follow an initial non-empty pattern to search, using `^` after a space to indicate "not" is viable, e.g. `PATT1 ^PATT2`.

I don't understand this. Why wouldn't the pattern consist solely of `NOT ^sth`? A single not filter is a valid usage.

> 3.  or `-` could be used (after a space) to indicate "not", e.g. `PATT1 -PATT2`.
> 4.  with 3) above, the "and" could also be formed with a `+` like so: `PATT1 +PATT2`. This allows a space to still fuzzy-match a space, i.e. match the space or something else.

`-` seems nice. `\-` will then be its escape? ` +` makes it more verbose to type the patterns, but is otherwise good. One point of caution though: Using these custom syntax will increase the entry barrier. People would already be familiar with the space as an AND operator, from web search engines and fzf.

**Comment by genivia-inc** [comment]
The Google search engine uses `OR`, `AND` and `-`, where `AND` is the default when terms are separated by spacing. The prefix `-` excludes terms and phrases. The prefix `+` forces an exact match of a term or phrase or terms and phrases can be quoted with `"` for exact matching. Brackets can be used to group `AND` and `OR` operators. It is likely that Google normalizes bracketed queries to CNF internally. Instead of `OR`, regex syntax offers `|` alternations. If users are not familiar with that operator (insert facepalm here) then we should consider `OR` or a comma `,` perhaps?

**Comment by genivia-inc** [comment]
We are getting off topic as this is no longer about fuzzy search, which is implemented. The AND/NOT logic in patterns and AND/NOT command-line options are a separate topic.

In the meantime an `--and` command-line option is in the works. Note that we already have `--not`.

Until the `--and` is added as an option there won't be progress on using AND/NOT logic within patterns. When the `--and` is done we can open a new discussion topic.

**Comment by genivia-inc** [comment]
It's been a while, but finally ugrep v3.1 is released with Boolean search with `--bool`, `--and` and `--not` options. The new `--bool` option specifies Boolean search queries as patterns with AND (same as spacing), OR (same as `|`), and NOT (same as `-`) operators, literally quoting with `" "`, and grouping with `(` and `)`. See [Boolean search queries with --bool (-%), --and, --not](https://github.com/Genivia/ugrep#bool) for details and examples. This works in the query UI, simply press Alt-% to activate the `--bool` mode. The `--bool` option can be used with `-F` to restrict the search to strings, and with `-Z` say `-Z+4` for fuzzy search (`-Z+4` to allow up to 4 extra chars). There are practically no limits on how to combine options to get the best results.

**Comment by NightMachinary** [comment]
[**@genivia-inc**](https://github.com/genivia-inc) commented on [Dec 14, 2020, 11:38 PM GMT+3:30](https://github.com/Genivia/ugrep/issues/31#issuecomment-744680678 "2020-12-14T20:08:48Z - Replied by Github Reply Comments"):
> It's been a while, but finally ugrep v3.1 is released with Boolean search with `--bool`, `--and` and `--not` options. The new `--bool` option specifies Boolean search queries as patterns with AND (same as spacing), OR (same as `|`), and NOT (same as `-`) operators, literally quoting with `" "`, and grouping with `(` and `)`. See [Boolean search queries with --bool (-%), --and, --not](https://github.com/Genivia/ugrep#bool) for details and examples. This works in the query UI, simply press Alt-% to activate the `--bool` mode. The `--bool` option can be used with `-F` to restrict the search to strings, and with `-Z` say `-Z+4` for fuzzy search (`-Z+4` to allow up to 4 extra chars). There are practically no limits on how to combine options to get the best results.

I somehow missed this update, and now that I have returned to see how ugrep has grown, it is really awesome! Thanks a lot!
